create_component.sh: add script for creating component boilerplate

maceksimon opened the pull request:
Adds a shell script for creating a component boilerplate.

This should streamline the creation of new components and ensure unified structure.

To create a component, run `./create_component.sh component_name component_folder`

tmszi commented:
Better prog. lang for this usecase is TypeSript because:

1. BASH shell interpreter is installed by default on the OS GNU/Linux, UNIX, but is missing in OS MS Windows.
2. Our app require Node.js to be installed on all OS platforms by default, which allow to run TypeScript without additional interpreter installation
3. You should write simple unit test with [Cypress unit test ](https://github.com/auto-mat/ride-to-work-by-bike-frontend/blob/main/test/cypress/e2e/unit_tests_app_code.spec.cy.js)(compare content of generated component file), please
4. File script path should be ./src/utils/create_component_file.ts

maceksimon commented:
@tmszi  the rewrite is ready, but the testing is complicated.
Node environment is set to use `require()` while the test environment uses module `import`.
In other words - exporting the functions from the script for testing makes the script fail when run in console.

tmszi commented:
> @tmszi the rewrite is ready, but the testing is complicated. Node environment is set to use `require()` while the test environment uses module `import`. In other words - exporting the functions from the script for testing makes the script fail when run in console.

I'm really wonder you don't use **TypeScript** (my comment above: _Better prog. lang for this usecase is TypeScript..._) by default (with using ES6 import module directive)?

Read this doc https://www.digitalocean.com/community/tutorials/how-to-use-modules-in-typescript, please.

1. If file script require other file scripts to be imported organize them inside directory e.g. ./src/utils/create_component_file/, please
2. Templates string code should be inside standalone files inside dir e.g. ./src/utils/create_component_file/templates/ and you read content files inside your code.

maceksimon commented:
Ok, you are right. I did not realize that this should be run through the compiler within the node process.
I updated the script and templates structure. I also added a unit test of the generated file.
Hopefully this is closer to what you imagined.

One change we need for it to work is to set the `watchForFileChanges: false` in E2E test suite, because otherwise the test keeps restarting as the new file is being created within in.

tmszi commented on test/cypress/e2e/unit_tests_app_code.spec.cy.js:
```diff
@@ -12,3 +14,61 @@ describe('Unit Test Application Code', function () {
+
+  after(() => {
+    // Optional cleanup
+    cy.exec(`rm -rf src/components/${componentFolder}/${componentName}.vue`);
```
`rm -rf` is not multi platform OS command.

tmszi commented on test/cypress/e2e/unit_tests_app_code.spec.cy.js:
```diff
@@ -12,3 +14,61 @@ describe('Unit Test Application Code', function () {
+
+  it('should create a Vue file in the specified directory', () => {
+    // Check that file exists
+    cy.exec(`test -f src/components/${componentFolder}/${componentName}.vue`)
```
`test -f ` is not multi platform OS command.

maceksimon replied:
Refactored into a Cypress task which uses `fs` to check and delete files.

tmszi commented:
> One change we need for it to work is to set the `watchForFileChanges: false` in E2E test suite, because otherwise the test keeps restarting as the new file is being created within in.

You can try this Cypress event https://docs.cypress.io/api/plugins/before-run-api.

tmszi commented on src/utils/create_component_file/index.ts:
```diff
@@ -0,0 +1,48 @@
+const fs = require('fs');
```
Missing  module doc header.

maceksimon replied:
Added doc header.

tmszi commented on src/utils/create_component_file/index.ts:
```diff
@@ -0,0 +1,48 @@
+const fs = require('fs');
```
Use ES6 import please.

maceksimon replied:
Rewritten to use ES6 import.

tmszi commented on src/utils/create_component_file/index.ts:
```diff
@@ -0,0 +1,48 @@
+
+// Ensure the component and tests directories exist
+const projectRoot = process.cwd();
+ensureDir(path.join(projectRoot, `src/components/${folderPath}`));
```
`path.existsSync("/the/path")`

tmszi commented on src/utils/create_component_file/templates/template_component.ts:
```diff
@@ -0,0 +1,42 @@
+export default `<script lang='ts'>
```
Template should be single TXT file template_component.txt not TS module.

tmszi commented on test/cypress/e2e/unit_tests_app_code.spec.cy.js:
```diff
@@ -12,3 +14,61 @@ describe('Unit Test Application Code', function () {
+  before(() => {
+    // Pre-test setup to ensure a clean state and run the script
+    cy.exec(
+      `npx tsc src/utils/create_component_file/index.ts && node src/utils/create_component_file/index.js ${componentName} ${componentFolder}`,
```
Instead of `node ...` use `npx` please.

maceksimon replied:
This is not a simple replacement - `npx` fails here, because it's function is to execute binaries from `node_modules`. What is the reason for not using node command?

tmszi commented on test/cypress/e2e/unit_tests_app_code.spec.cy.js:
```diff
@@ -12,3 +14,61 @@ describe('Unit Test Application Code', function () {
+
+  it('should create a Cypress test file in the correct directory', () => {
+    // Check that file exists
+    cy.exec(`test -f src/components/__tests__/${componentName}.cy.js`)
```
`test -f`  is not multi platform OS command.

maceksimon replied:
Refactored into a Cypress task which uses `fs` to check and delete files.

tmszi commented on test/cypress/e2e/unit_tests_app_code.spec.cy.js:
```diff
@@ -12,3 +14,61 @@ describe('Unit Test Application Code', function () {
+
+  before(() => {
+    // Pre-test setup to ensure a clean state and run the script
+    cy.exec(
```
`&&` is mutli platform OS behavior?

maceksimon replied:
Yes, this works in [cmd shell](https://ss64.com/nt/syntax-conditional.html) and in [Unix bash and zsh](https://www.gnu.org/software/bash/manual/bash.html#Definitions).

tmszi commented on src/utils/create_component_file/index.js:
```diff
@@ -0,0 +1,51 @@
+'use strict';
```
This file should not be a part of Git repository, it is generated dynamically.

maceksimon replied:
Removed (added to .gitignore)

tmszi commented on package.json:
```diff
@@ -53,7 +53,8 @@
     "test:component:open:win:edge": "cypress open --component --browser edge",
-    "prepare": "husky install"
+    "prepare": "husky install",
+    "create:component": "npx tsc src/utils/create_component_file/index.ts && node src/utils/create_component_file/index.js"
```
`&&` is multi platform behavior?

maceksimon replied:
Yes, this works in [cmd shell](https://ss64.com/nt/syntax-conditional.html) and in [Unix bash and zsh](https://www.gnu.org/software/bash/manual/bash.html#Definitions).

tmszi commented on src/utils/create_component_file/index.ts:
```diff
@@ -0,0 +1,48 @@
+
+// Ensure the component and tests directories exist
+const projectRoot = process.cwd();
+ensureDir(path.join(projectRoot, `src/components/${folderPath}`));
```
Test should create own temporary directory which store dynamically generated file structure e.g. `/tmp/tmpfqq9piuk/src/components/${folderPath}` and delete automatically after test finish.

maceksimon replied:
I understand. However, is it not also valuable to test that files are being created in the correct directory within the file structure?

tmszi replied:
Yes it is. For testing purpose you change root dir only (instead of dir ride-to-work-by-bike-frontend/src/components you can use e.g. /tmp/tmpfqq9piuk/src/components/), but file structure is the same.

Create custom temp dir with Node.js

https://blog.mastykarz.nl/create-temp-directory-app-node-js/

tmszi replied:
But yes , is one of another option, if you use real root dir, but is important to preserved `watchForFileChanges: true`.

maceksimon replied:
I ended up running the test with `watchForFileChanges: true` without additional changes. It seems that the rerunning of the script in the earlier version of the test was caused by something else.

maceksimon commented on src/utils/create_component_file/templates/template_test.ts:
```diff
@@ -0,0 +1,37 @@
+export default `import COMPONENT_NAME from 'IMPORT_PATH.vue';
```
Changed to .txt file.
